src/nts_ke/server/connection.rs

```diff
@@ -199,6 +227,61 @@ impl KeServerConn {
             }
 
             let keys = gen_key(&self.tls_session).unwrap();
+            let mut buf = &buf[..];
+            while self.ntske_state.finished == false {
+                // need to read 4 bytes to get the header.
+                let mut header: [u8; HEADER_SIZE] = [0; HEADER_SIZE];
+                if let Err(error) = buf.read_exact(&mut header) {
+                    error!(self.logger, "read nts-ke record header: {}", error);
+                    self.shutdown();
+                    return;
+                }
+
+                // need to read the body_length to get the body.
+                let body_length = u16::from_be_bytes([header[2], header[3]]);
+                let mut body = vec![0; body_length as usize];
+                if let Err(error) = buf.read_exact(&mut body) {
```

Comment:
This is a nonblocking socket, so may very well get an EAGAIN. read_exact doesn't handle this well.

Reply:
Fixed it using buffer.

Reply:
So now there is another problem, namely that you reparse over and over again. Well, that's probably fine. More concerning is that it isn't clear what length the buffer can maximally get to. Also we need to ignore unknown, non-critical records. Otherwise there is an extensibility issue: the protocol can't evolve if unknown records trigger errors.

This is definitely a step in the right direction, and thanks for being willing to take this one on, but it does involve a bunch of complex issues. There is a reason I had punted on it earlier.

Reply:
> namely that you reparse over and over again.

 Modified that when stream is not enough to read NTS-KE record, update ntske_buffer with reading buffer.
`nts_ke/server` is updated not to  reparse over and over again.

Reply:
> Also we need to ignore unknown, non-critical records.

Fixed to ignore DeserializeError::UnknownNotCriticalRecord.